src/cargo/core/compiler/job_queue.rs

```diff
@@ -36,13 +86,49 @@ pub struct JobQueue<'a, 'cfg> {
     queue: DependencyQueue<Unit<'a>, Artifact, Job>,
     tx: Sender<Message>,
     rx: Receiver<Message>,
-    active: HashMap<u32, Unit<'a>>,
+    active: HashMap<JobId, Unit<'a>>,
     compiled: HashSet<PackageId>,
     documented: HashSet<PackageId>,
     counts: HashMap<PackageId, usize>,
     progress: Progress<'cfg>,
     next_id: u32,
     timings: Timings<'a, 'cfg>,
+
+    /// These are tokens that we have acquried but aren't actively using.
+    ///
+    /// We should only have tokens here "in-between" loops of the
+    /// `drain_the_queue` function, i.e., when we've acquired a bunch of tokens
+    /// and are about to give them to rustcs.
+    unused_tokens: Vec<Acquired>,
+
+    /// This keep track of the in-use tokens for each rustc process.
+    ///
+    /// If we're operating in jobserver-per-rustc mode, then this is going to
+    /// contain the number of tokens each rustc has. Otherwise, it's going to
+    /// contain just the one token.
+    ///
+    /// If we ever wind up in a situation where an entry in this map is empty,
+    /// then something has gone wrong, as we have a rustc running without tokens
+    /// associated with it.
+    rustc_tokens: HashMap<JobId, Vec<Acquired>>,
```

Comment:
I think we're going to want to avoid this structure to handle the implicit token, we I believe can get away with the tokens in this map are purely for "extra parallelism" for each rustc. There'd be a separate vector, whose length is less than one from this map, which has all the tokens used to spawn the rustc process in the first place.